rpc/src/client.rs

```diff
@@ -1,45 +1,150 @@
+//! you need more configuration options and / or some custom client you can create
+//! one using [`jsonrpsee`] crate directly.
 
 #[cfg(not(target_arch = "wasm32"))]
-mod imp {
-    use crate::Result;
+pub use self::native::Client;
+
+#[cfg(not(target_arch = "wasm32"))]
+mod native {
+    use std::fmt;
+    use std::result::Result as StdResult;
+
+    use crate::{Error, Result};
+    use async_trait::async_trait;
     use http::{header, HeaderValue};
+    use jsonrpsee::core::client::{BatchResponse, ClientT, Subscription, SubscriptionClientT};
+    use jsonrpsee::core::params::BatchRequestBuilder;
+    use jsonrpsee::core::traits::ToRpcParams;
+    use jsonrpsee::core::Error as JrpcError;
     use jsonrpsee::http_client::{HeaderMap, HttpClient, HttpClientBuilder};
     use jsonrpsee::ws_client::{WsClient, WsClientBuilder};
+    use serde::de::DeserializeOwned;
 
-    pub fn new_http(conn_str: &str, auth_token: Option<&str>) -> Result<HttpClient> {
-        let mut headers = HeaderMap::new();
+    /// Json RPC client.
+    pub enum Client {
+        Http(HttpClient),
+        Ws(WsClient),
+    }
 
-        if let Some(token) = auth_token {
-            let val = HeaderValue::from_str(&format!("Bearer {token}"))?;
-            headers.insert(header::AUTHORIZATION, val);
-        }
+    impl Client {
+        /// Create a new Json RPC client.
+        ///
+        /// Only 'http' and 'ws' protocols are supported and they should
+        /// be specified in the provided `conn_str`. For more flexibility
+        /// consider creating the client using [`jsonrpsee`] directly.
+        pub async fn new(conn_str: &str, auth_token: Option<&str>) -> Result<Self> {
+            let mut headers = HeaderMap::new();
+
+            if let Some(token) = auth_token {
+                let val = HeaderValue::from_str(&format!("Bearer {token}"))?;
+                headers.insert(header::AUTHORIZATION, val);
+            }
 
-        let client = HttpClientBuilder::default()
-            .set_headers(headers)
-            .build(conn_str)?;
+            let client = match conn_str.split_once(':') {
+                Some(("http", _)) => Client::Http(
+                    HttpClientBuilder::default()
+                        .set_headers(headers)
+                        .build(conn_str)?,
+                ),
+                Some(("ws", _)) => Client::Ws(
+                    WsClientBuilder::default()
+                        .set_headers(headers)
+                        .build(conn_str)
+                        .await?,
+                ),
+                _ => return Err(Error::ProtocolNotSupported(conn_str.into())),
+            };
 
-        Ok(client)
+            Ok(client)
+        }
     }
 
-    pub async fn new_websocket(conn_str: &str, auth_token: Option<&str>) -> Result<WsClient> {
-        let mut headers = HeaderMap::new();
+    #[async_trait]
+    impl ClientT for Client {
```

Comment:
I thought we wanted to hide anything related to `jsonrpsee`. Am I missing something?

Reply:
yes, so the #100 was about having to manually include `jsonrpsee` in dependencies for typing the client we returned from `celestia_rpc::client::` helpers. We still want to have all the `(Blob|Share|Header|P2p|etc)Client` in our API and allow for interacting with them through custom `jsonrpsee` clients I think.

If we were to hide the `jsonrpsee` completely then we'd need to write a boilerplate for all the rpc modules in a `Client` and also provide more flexible interface allowing eg. setting max_request / response sizes or timeouts?

Reply:
Ok got it.